### DISCUSSION TITLE
actions/runner nodejs plan and next steps

### DISCUSSION THREAD

**Discussion by benwells** [community/community#160454] [Actions]
👋 Hey all!

Sorry we've been quiet on this.  I want to provide you with as much clarity as we have right now on the next steps for node.

### Node 12

We've been on the path to deprecate this for a while, we started back in October with a deprecation message.  We are super relieved 😅 that we will be [removing Node12 from the Actions runner on the 14th of August 2023](https://github.blog/changelog/2023-07-17-github-actions-removal-of-node12-from-the-actions-runner/).  This has taken a while because a bunch of folks were still actively using node12 and we wanted to give them heads up and migrate as many of them as we could, in advance.

### Node 16 and 20

From here, we plan to skip over node 18 and hop on the node 20 LTS release.  We are currently working on a date to rollout node 20 and have work in-flight to accomplish this.  We will update this discussion in the coming weeks with our plans once we finalize them.

This does mean that we will carry node 16 past it's out-of-support date, but will start the deprecation process of node 16 later this year, similar to how we deprecated node 12 over a period of time.

Thank you all for your engagement and feedback on this matter!  Please keep raising issues in our repo when you have concerns or feedback.  We really appreciate the engagement and support ❤️ .

**Comment by wdzeng** [comment]
Happy to hear that GitHub finally had response on this.

**Comment by AWare** [comment]
it's really disappointing that this is has only been communicated this recently, and that most issues relating to this have been closed for months.

https://github.com/actions/runner/issues/2255
https://github.com/actions/github-script/issues/310

It's also disappointing that github can both not support any later version than 16 in an action, while demanding a minimum version of 18 to use octokit. https://github.com/octokit/octokit.js/#fetch-missing 

Please can this be reconsidered?

**Reply by AWare** [reply]
I have also raised an issue in octokit https://github.com/octokit/octokit.js/issues/2505

**Reply by nebuk89** [reply]
Hey! Really sorry this has come about :( I just want to clarify, if we are able to accelerate a move to node20 would this solve this for you? (our preference is still to get as far ahead as we can as we know it's Actions that has lagged here 😞 )

**Reply by AWare** [reply]
That would certainly be appreciated thanks, but I think more important would be coordination between the various teams to keep a consistent set of supported node versions.

**Reply by AWare** [reply]
@nebuk89 thanks for your accelerated release of node 20.

**Comment by kuvaus** [comment]
This is great to hear! Thanks a lot!

One small suggestion:

It took me a bit of googling and browsing various github actions issues to find this post.
Since most github actions depend on `Node.js`, it might be worth to somehow pin the original post or maybe make a post on the [blog](https://github.blog/tag/github-actions/) so that people immediately see what are Github's official plans on this. It is relieving to see that when `Node 20 LTS` comes, all the action developers can just smoothly switch to that. Maybe you're already planning a bigger official post when you can finalize the exact date.


As for https://github.com/actions/runner/discussions/2704#discussioncomment-6530252:

For now you can either use `Node 16`, import `node-fetch` and give that to `Octokit`:
```javascript
const fetch = require('node-fetch').default;
const octokit = new Octokit({ auth: token, request: {fetch: fetch, } });
```


Or you can make a composite action with `Node 18`:

With composite actions in `action.yml` you need to also give the inputs as environment variables to your `index.js`:

```yaml
inputs:
  my_parameter:
    description: 'My input'
    required: false
```
And after setting up `Node 18`

```yaml
      env:
        INPUT_MY_PARAMETER:          ${{ inputs.my_parameter }}
```      
If you need, I have those in more detail in [Node16](https://github.com/kuvaus/changelog-releasenotes-action/blob/node16/src/index.js) and [Node18](https://github.com/kuvaus/changelog-releasenotes-action/blob/node18/action.yml) branches of one of my actions.


Now, with the upcoming support for `Node 20` I can remove both of these workarounds. This makes the code much simpler which is always lovely. :)

**Reply by AWare** [reply]
while this is a very helpful comment and workaround, I don't think this is a reasonable expectation on githubs part.

**Reply by DariuszPorowski** [reply]
@AWare Did PoC with workaround a while ago, see: https://github.com/orgs/community/discussions/53217#discussioncomment-6227341
I think the approach is similar to @kuvaus

**Comment by razor-x** [comment]
Since [Node 20 won't be LTS until `2022-10-24`](https://nodejs.dev/en/about/releases/) does that mean the soonest we can expect support is around early November?

Can you share any updates on if your team is making any process changes to ensure this doesn't happen in the future? Thanks.

**Comment by ChristopherHX** [comment]
Node20 support has been merged <https://github.com/actions/runner/pull/2732> by the runner team

It is now only a matter of waiting for the next runner release, if nobody reverts it
Maybe in 14th August? node12 has been removed already from the default branch. Yes I'm speculating.

**Reply by MikeMcC399** [reply]
@ChristopherHX Thanks for the heads-up! I couldn't see any documentation PR in the pipeline, but maybe `node20` will get released silently first as a pre-release?

**Reply by ChristopherHX** [reply]
@MikeMcC399 You should not make use of node20 until 14th September if released on 14th August.

- Self-Hosted runners have a 1 month grace period on github.com to upgrade manually
- GHES users won't be able to use node20 for a longer time, without manual upgrade of all self-hosted runners without an official backport
- Create a new major release / tag

So yes it's better to wait longer.

Azure Devops allowed to specify node16 + node20 (prefer newer node) to be backward compatible, while GitHub decided removing this feature(while forking the agent/runner in 2019, this was never possible it Github Actions).

**Reply by MikeMcC399** [reply]
> Create a new major release / tag

Absolutely! This is a breaking change and needs a new major release version number for any published action using it.

**Comment by MikeMcC399** [comment]
@benwells 

Do you have any update for us on the release date for Node.js `20` support in JavaScript Actions? This would help very much for downstream planning.

**Reply by MikeMcC399** [reply]
My question was answered in https://github.com/actions/runner/issues/2619#issuecomment-1679003443

> Hello, latest runner release ([2.308.0](https://github.com/actions/runner/releases/tag/v2.308.0)) contains latest node16 version and node20 version. This runner version will be fully rolled-out by the end of this week. After release you will be able to use `'node20'` actions

**Reply by MikeMcC399** [reply]
I have now successfully used `node20` with runner version ([2.308.0](https://github.com/actions/runner/releases/tag/v2.308.0)) 👍🏻

**Comment by MikeMcC399** [comment]
@benwells 

- Since you've now released `node20` support, I ask for it to be added to the documentation. See https://github.com/actions/runner/issues/2786.

**Reply by MikeMcC399** [reply]
- https://github.com/actions/runner/issues/2786 in this repo was rejected. I refiled under https://github.com/github/docs/issues/27673 (this now says "This issue has been moved to a repository you don't have access to.", which means it is being worked on, but progress is not visible for external users). 
- [`runs.using` for JavaScript actions](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsusing-for-javascript-actions) to read current contents.

**Reply by MikeMcC399** [reply]
Documentation [`runs.using` for JavaScript actions](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsusing-for-javascript-actions) is now fixed as refers only to `node20`.

**Comment by MikeMcC399** [comment]
@benwells 

In your original post in July 2023 you wrote:

>  We will update this discussion in the coming weeks with our plans once we finalize them.

That is currently the first and last post in this discussion from any maintainer. The milestone of Node.js `16` end-of-life on Sep 11, 2023 has now been reached, so it would be highly appreciated to hear of your further detailed planning!

- [x] Can we assume that `node20` support will be announced in https://github.blog/changelog/ soon?

- [x] What will happen with `node16`? Will it be deprecated?

- [x] Will GitHub JavaScript Actions specifying `node16` be forced to run on `node20`?

- [x] Will GitHub JavaScript Actions specifying `node12` stop working, continue to be forced on to `node16` or will they be forced to run on `node20`?

**Reply by MikeMcC399** [reply]
The list of questions is mostly answered in https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/

The only unanswered question is:

- [ ] Will GitHub JavaScript Actions specifying `node12` stop working, continue to be forced on to `node16` or will they be forced to run on `node20`?

**Reply by thboop** [reply]
Eventually you will be forced onto to the "lowest node version we support" with an appropriate warning so you know it is happening. I'll make sure it is clear in future communications, thanks for the feedback @MikeMcC399

**Reply by MikeMcC399** [reply]
@thboop 

Thanks very much for explaining about what will happen with older versions. That makes absolute sense!

One more feedback point: [GitHub Actions: Transitioning from Node 16 to Node 20](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/) Sep 22, 2023 announced "Our plan is to transition all actions to run on Node 20 by Spring 2024."

Using northern hemisphere seasons is not a globally compatible way of naming time-frames. Spring 2024 is 6 months later in the southern hemisphere compared to the northern hemisphere. If you use months (or month ranges like Apr - Jun) or calendar quarters, like second calendar quarter 2024 for instance, that would be correct over the whole globe.

**Comment by gilzow** [comment]
The blog post states
> Given this, we will remove Node12 from the Actions runner on the 14th of August 2023

But the actions platform still states Node12 as a valid option if you use the wrong argument for the `runs:using` property

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''using: nod16' is not supported, use 'docker', 'node12', 'node16' or 'node20' instead.')
```

**Reply by MikeMcC399** [reply]
@gilzow 

If a runner specifies `node12` then it currently runs as `node16` and the message is:


> The following actions uses node12 which is deprecated and will be forced to run on node16: <name-of-action-specifying-node12>. For more info: https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default

So it can still be specified as a parameter but `node12` isn't actually there anymore.

Does that make sense?

**Reply by muzimuzhi** [reply]
> But the actions platform still states Node12 as a valid option if you use the wrong argument for the `runs:using` property
> 
> ```
> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''using: nod16' is not supported, use 'docker', 'node12', 'node16' or 'node20' instead.')
> ```

BTW, this exception is thrown by [`main/src/Runner.Worker/ActionManifestManager.cs`](https://github.com/actions/runner/blob/main/src/Runner.Worker/ActionManifestManager.cs) in https://github.com/actions/runner repo.

**Reply by MikeMcC399** [reply]
- https://github.com/actions/runner/pull/2717 removed `node12` as part of release [v2.308.0](https://github.com/actions/runner/releases/tag/v2.308.0) 
- https://github.com/actions/runner/issues/2619#issuecomment-1679003443 notes on Aug 15, 2023 that this release would be rolled out in that week

**Comment by MikeMcC399** [comment]
According to the blog post [GitHub Actions: Transitioning from Node 16 to Node 20](https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/) Sep 22, 2023 I would expect to see warnings now for `node16` usage (since we are past the date Oct 23, 2023), but I don't see any additional warnings on GitHub-hosted runners such as `ubuntu-22.04`.

> Node 16 has reached its [end of life](https://github.com/nodejs/Release/#end-of-life-releases), prompting us to initiate its deprecation process for GitHub Actions. Our plan is to transition all actions to run on Node 20 by Spring 2024. We will actively monitor the migration's progress and gather community feedback before finalizing the transition date. Starting October 23rd, workflows containing actions running on Node 16 will display a warning to alert users about the upcoming migration.

Has there been a change or delay to the plan?

**Reply by muzimuzhi** [reply]
I can already see such warnings in a [recent run](https://github.com/muzimuzhi/latex3/actions/runs/6630737668).

> <job name>
> Node.js 16 actions are deprecated. Please update the following actions to use Node.js 20: actions/cache/restore@v3, actions/cache/save@v3, actions/upload-artifact@v3. For more information see: https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/.

**Reply by MikeMcC399** [reply]
@muzimuzhi 

Thanks for sharing! It seems to be a staged-rollout in that case. I just ran a workflow which contained `actions/upload-artifact@v3` and I did not receive any warning.

**Reply by MikeMcC399** [reply]
@muzimuzhi 

Thanks again! I am now receiving the deprecation warnings as expected. 🙂

**Reply by MikeMcC399** [reply]
- ... and now the warnings have been disabled - see https://github.com/actions/upload-artifact/issues/444#issuecomment-1779785321

**Comment by Danielku15** [comment]
Can someone shed some light on the plan with GHES? We are on GHES 3.10.3 and the latest runner provided there is 2.304.0 (indicated in the 'New Runner' page and also auto-upgrade). But for node20 we need 2.308.0. 

GHES 3.11 is "only" in RC and also typically our internal IT does not upgrade directly. 

As actions in the market are moving to node20 things start to get tricky as we need bugfixes only provided on new versions already targeting node20.

**Comment by MikeMcC399** [comment]
According to https://github.com/actions/upload-artifact/issues/444#issuecomment-1779785321 from @joshmgross

> 👋 We've disabled the deprecation warning for now until we can get all of our official actions onto Node 20.

As far as I can see, the last related blog entry is https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20/ from Sep 22, 2023 and this says:

> Starting October 23rd (2023), workflows containing actions running on Node 16 will display a warning to alert users about the upcoming migration.

which has no mention about the warning having been disabled.

I couldn't find any related update in https://github.com/orgs/community/discussions/categories/actions either.

---

 **What is the plan for re-enabling the `node16` warnings?**

**Reply by joshmgross** [reply]
👋 Hey @MikeMcC399, thanks for the ping. I'll find someone who can provide at timeline.

**Reply by thboop** [reply]
Hey @MikeMcC399 , we want to heavily encourage upgrading to node 20. The warnings will be coming back on in the next couple of weeks, alongside the full removal of node 16 in spring 2024.

Node 16 workflows at that time will then automatically run on node 20, with a warning to users to let them know.

**Reply by MikeMcC399** [reply]
Hi @thboop 

Thank you very much for sharing the plans! That is very helpful. 👍🏻

**Reply by MikeMcC399** [reply]
I'm seeing the warning on GitHub-hosted runners appearing again.

**Comment by MikeMcC399** [comment]
There is a new GitHub blog post [GitHub Actions; All Actions will run on Node20 instead of Node16 by default](https://github.blog/changelog/2024-03-07-github-actions-all-actions-will-run-on-node20-instead-of-node16-by-default/) announcing that all actions will be forced to run under Node.js `20` starting on May 13, 2024.

**Comment by ScottBrenner** [comment]
Hello, found this discussion while searching around a bit - is support for Node 22 on the roadmap?

https://github.com/actions/setup-node added support for Node 22 as of https://github.com/actions/node-versions/releases/tag/22.0.0-8879734543, yet the runner errors when configured to use Node 22 in https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsusing-for-javascript-actions:

> `Error: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter ''using: node22' is not supported, use 'docker', 'node12', 'node16' or 'node20' instead.')`

via https://github.com/jaymzh/main-branch-check-action/pull/1#issuecomment-2123493663

**Reply by muzimuzhi** [reply]
I think they are different things.
- use of `runs.using: 'node22'` in metadata file (`action.y(a)ml`) for JavaScript actions is not supported yet.
- use of action
  ```yaml
  - uses: actions/setup-node@v4
    with:
      node-version: 22
  # more steps using node and/or npm
  ```
  is supported, though not revealed from the doc of `actions/setup-node@v4`.

**Reply by bmuenzenmeyer** [reply]
Ran into this today - it's a pretty significant use case, IMO, in that it would support actions to be written in TypeScript and use `--experimental-strip-types` to run them natively.

**Reply by iamstarkov** [reply]
@muzimuzhi is this a relevant repo/issue to discuss support for `runs.using: node22`?

**Reply by iamstarkov** [reply]
relevant issue seems to be neglected https://github.com/actions/runner/issues/3600

**Reply by MikeMcC399** [reply]
- In the past GitHub Actions has not attempted to cover each Node.js LTS version for `runs.using`. Based on the past strategy there would be a jump from `20` to `24` when Node.js 24 is released without covering `22`. Of course the GitHub Actions team should be the ones providing the answers here and if they replied to #3600 then the situation could be clarified.

**Reply by iamstarkov** [reply]
@MikeMcC399 fair enough, I've asked people who've driven nodejs22 support https://github.com/actions/runner/pull/2732#issuecomment-2619045154

**Comment by nebuk89** [comment]
Sorry for the slowness on what comes next after node20! I was out for a few months but now I am back I will pick up with the team and start the annual wheel turn of moving up our node versions :D

**Reply by iamstarkov** [reply]
@nebuk89 any issue or pr to follow for status updates?

**Reply by MikeMcC399** [reply]
@nebuk89 

Do you have any update about a follow-on for `node20`?
- Would you also like to respond to issue https://github.com/actions/runner/issues/3600

**Reply by MikeMcC399** [reply]
@nebuk89 

Asking again. Are you planning to support Node.js 22 or will you skip and support Node.js 24 after it is released and then promoted later this year (hopefully) to LTS status.

**Comment by cyberpvn7** [comment]
ok

**Comment by nebuk89** [comment]
We plan to jump straight to node24 as it will be an LTS. We are still working on when we will start this upgrade :) but it will be some time in the next few months. Thanks for keeping engaged <3

**Reply by iamstarkov** [reply]
@nebuk89 what about node26? is it planned to be skipped as well?

**Reply by MikeMcC399** [reply]
This was announced in [Deprecation of Node 20 on GitHub Actions runners](https://github.blog/changelog/2025-09-19-deprecation-of-node-20-on-github-actions-runners/)
